spi: stm32: Enable DMA client

FRASTM opened the pull request:
to support the spi as a client of the dma
This is an example spi_loopback of using dma with the spi client.

The DT is filled with parameters to set the peripheral requests.
The dma channels used for the client transfer (rx, tx) are assigned by the DT among the total nb of dma channels (ex.7 per dma instance)

This PR is tested using the nucleo_f411re and nucleo_l476rg

This PR could also run through dmamux, thanks to https://github.com/zephyrproject-rtos/zephyr/pull/22021 on nucleo_wg55rg for example, in PR https://github.com/zephyrproject-rtos/zephyr/pull/23155

This PR is required to fix https://github.com/zephyrproject-rtos/zephyr/issues/20825

Signed-off-by: Francois Ramu francois.ramu@st.com

EDIT by avisconti::
In order to build it properly I had to use PERIPHERAL and MEMORY, see below).
After that I ran w/o problems the microphone sample on shield iks02a1
(see https://github.com/avisconti/zephyr/tree/test_dma)
EDIT by erwango:: This issue has been fixed.

tbursztyka commented:
Don't open 2 PR. continue the work in #23155  And when ready just edit it's name to remove the [wip].

FRASTM commented:
fix Nits issues and Kconfig issues

FRASTM commented:
add commit to initialize the dma peripheral request in DMA V2

FRASTM commented:
- Removing the reference to mux
- In the test/drivers/spi_loopback application :
 when using a DMA, the testcases which are sending from a null tx buffer does not sent NOP on the MOSI line, but sends actual data from @ 0x0. Should we consider it is an error, from the spi driver in dma mode or from the dma driver ?
 In dma mode, skipping 'half end' and 'every 4' test cases (purpose ?)

FRASTM commented:
With DMA, the spi driver is changed to send NOP on the line in case the Tx buff is null
Then it acts like the fifo mode. THe source buffer is replaced by the nop char, without increment.

FRASTM commented:
the Tx buff test case passed

FRASTM commented:
avoid changes on dma-controller.yaml

FRASTM commented:
remove dma_stop on transfer completion

FRASTM commented:
merge transceive_dma functions into transceive function

FRASTM committed:
dts: arm: stm32l4: configure dma instances

This patch is defining the DMA1 & 2 for the sm32l4 series
with 7 channels and support of mem2mem transfers

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Adapt tests/drivers/dma applications to run with STM32
Configure test for running on nucleo_l476 board

FRASTM commented:
change spi driver to support dma and non-dma modes on several SPI instances

FRASTM commented:
set a macro for `stm32_dma_set_channel` incase of DMA V1

erwango commented on include/dt-bindings/dma/stm32_dma.h:
```diff
@@ -1,16 +1,44 @@
+#define ZEPHYR_INCLUDE_DT_BINDINGS_DMA_STM32_DMA_H_
+
 /* macros for channel-config */
+#if defined(CONFIG_DMA_STM32_V1)
```
Do we need to differentiate V1 vs V2?
There's no difference made in "st,stm32-dma.yaml " binding

FRASTM replied:
the only diff is about the fifo threshold for V1 (V2 has not)

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -46,14 +49,181 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
+		&& (data->dma_rx.transfer_complete != true));
+#else
 	return spi_context_tx_on(&data->ctx) || spi_context_rx_on(&data->ctx);
+#endif /* CONFIG_SPI_STM32_DMA */
```
This won't work when DMA and non DMA operate in same binary

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -46,14 +49,181 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
+#endif /* CONFIG_SPI_STM32_DMA */
 }
 
+#ifndef CONFIG_SPI_STM32_DMA
```
This won't work when DMA and non DMA operate in same binary

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -173,6 +344,9 @@ static void spi_stm32_shift_s(SPI_TypeDef *spi, struct spi_stm32_data *data)
  */
 static int spi_stm32_shift_frames(SPI_TypeDef *spi, struct spi_stm32_data *data)
 {
+#ifdef CONFIG_SPI_STM32_DMA
```
This won't work when DMA and non DMA operate in same binary

FRASTM replied:
what is not supported is spi1 Tx with dma and spi1 Rx without dma or  spi1 Rx with dma and spi1 Tx without dma

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -202,6 +377,7 @@ static void spi_stm32_complete(struct spi_stm32_data *data, SPI_TypeDef *spi,
 	}
 #endif
 
+#ifndef CONFIG_SPI_STM32_DMA
```
This won't work when DMA and non DMA operate in same binary

erwango commented on tests/drivers/spi/spi_loopback/boards/nucleo_l476rg.overlay:
```diff
@@ -0,0 +1,20 @@
+&spi1 {
+	compatible = "st,stm32-spi";
+	dmas = <&dma1 3 1 0x20440 0x04
+				&dma1 2 1 0x20480 0x04>;
```
Fix indentation

FRASTM replied:
done

erwango commented on tests/drivers/dma/chan_blen_loop_transfer/src/test_dma.c:
```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2016 Intel Corporation
+ *
```
Add your copyright

FRASTM replied:
changed

erwango commented on tests/drivers/dma/chan_blen_loop_transfer/src/main.c:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2016 Intel Corporation
+ *
```
Add your copyright

FRASTM replied:
changed

erwango reviewed (changes requested):
Commit "drivers: spi: enable stm32 spi driver for non-dma and dma modes" should be squashed with previous one.

FRASTM commented:
reduce size of this PR (with less commits) to enable spi client on DMA

FRASTM committed:
tests: drivers: spi loopback on nucleo_f411 board

This configures the nucleo_f411re board for testing the SPI loopback
with DMA V1 transfer. DMA instance 2 is used.
This test requires pin SPI_A_MISO (D12) to be connected
to pin SPI_A_MOSI (D11) on the nucleo_wb55rg boardtest dma_V1

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
merge bindings for dma V1 and V2

FRASTM commented:
add a configuratio for the Fifo mode in case of dma V1

FRASTM commented:
add one commit 37949eb9eab19a95cb366a799948c7d96dffca60 for irq

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -49,6 +61,22 @@ static void dma_stm32_irq_handler(void *arg)
+		func_ll_clear_ht[id](dma);
+		return;
+	}
+#else
```
now put in another commit about IRQ handling 37949eb9eab19a95cb366a799948c7d96dffca60

FRASTM committed:
tests: drivers: spi loopback on nucleo_l476 board

This configures the nucleo_l476rg board for testing the SPI loopback
when with DMA transfer.
It requires pin MOSI to be connected to pin MISO of the SPI1
SPI tx/rx are assigned to the request ID 1 of dma1 channels 3 and 2

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
change copyright in tests/drivers/dma/chan_blen_loop_transfer c-files

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -265,6 +309,14 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
+
+	if (&config->head_block->dest_address == NULL) {
+		LOG_INF("dest_buffer address is null.");
+	}
```
done in commit drivers: dma: control of null buffer address for stm32 dma transfers 5bf56cb9506aeb5682bf6d31962e98efa8335374

FRASTM commented:
rebase

tbursztyka commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -529,11 +544,14 @@ void test_spi_loopback(void)
 	    spi_rx_half_start(spi_fast, &spi_cfg_fast) ||
 	    spi_rx_half_end(spi_fast, &spi_cfg_fast) ||
 	    spi_rx_every_4(spi_fast, &spi_cfg_fast)
+
```
useless new line

FRASTM replied:
removed

tbursztyka commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -277,7 +280,12 @@ static int spi_rx_half_end(struct device *dev, struct spi_config *spi_conf)
 
-	LOG_INF("Start");
+	if (IS_ENABLED(CONFIG_SPI_STM32_DMA)) {
+		LOG_INF("Skip half end");
```
why?

FRASTM replied:
This test makes sens when transferring with Fifo. In case of DMA, the transfer is started for a defined length in a complete burst. The half transfer is possibly detected for the 1st part. But it has not been enabled here. The second half is then detected when the complete transfer ends.

tbursztyka replied:
ok

tbursztyka commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -340,7 +348,12 @@ static int spi_rx_every_4(struct device *dev, struct spi_config *spi_conf)
 
-	LOG_INF("Start");
+	if (IS_ENABLED(CONFIG_SPI_STM32_DMA)) {
+		LOG_INF("Skip every 4");
```
why?

FRASTM replied:
This test makes sens when transferring with Fifo. In case of DMA, the transfer is started for a defined length in a complete burst.

tbursztyka replied:
ok

tbursztyka commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -33,6 +33,8 @@ struct spi_cs_control spi_cs = {
 #define CS_CTRL_GPIO_DRV_NAME ""
 #endif
 
+/* connect MOSI pin and the MISO of the SPI */
```
Your commit message is misleading, it seems it just adds exception when stm32 dma is in use.

Also, this comment, and all the log message changes are not related to this commit, so have a separate patch.

FRASTM replied:
This is now splitted in 2 commits :
1. b972c68c0a89041c84eb18ba20a7fab32d2a8949 is for adding more comments to the test applications 

2. e75fc5ee50904ffec513330f637526c1d4194ba3 is to adapt for dma transfer

tbursztyka commented on tests/drivers/dma/loop_transfer/src/dma.c:
```diff
@@ -86,39 +103,43 @@ void main(void)
-		return;
-	}
+	for (chan_id = FIRST_DMA_CHANNEL;
+		chan_id <= LAST_DMA_CHANNEL;
```
indentation issue (should be below last '(' + 1 space)

FRASTM replied:
done

tbursztyka commented on tests/drivers/dma/chan_blen_transfer/src/main.c:
```diff
@@ -14,35 +14,27 @@
 #include <zephyr.h>
 #include <ztest.h>
 
-extern void test_dma_m2m_chan0_burst8(void);
```
Commit message is again not true. It is doing more than adapting the test for stm32 dma. Split the patch in 2, one that introduce the generic changes, one for the stm32 dma

FRASTM replied:
changed into different commits 17c3e68416dda8fa668c4f90110516f090651257 and eacaa587d30892a0aaa4e15ae04b9d8b867139e8

tbursztyka commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -449,11 +632,72 @@ static int transceive(struct device *dev,
+	}
+
+	if (asynchronous) {
+		return -ENOTSUP;
```
Would be nice to support it, the non-dma part supports it already. I think that's trivial to support.

FRASTM replied:
Same as `transceive` function
  @tbursztyka  What could be the asynchronous mode with a DMA transfer ?

tbursztyka replied:
It's like interrupt mode, you notify the caller through its k_poll_signal the transaction is done (or got some error).
You have set a dma_callback above that can be used for that I think.
So if asynchronous mode is requested by the user, you just don't run the loop at line 667 nor cs control nor release: you do these I believe in the dma cb.

But it's ok you could add this later in another PR.

tbursztyka commented on drivers/dma/dma_stm32.c:
```diff
@@ -196,20 +211,26 @@ static int dma_stm32_get_periph_increment(enum dma_addr_adj increment,
 					dev->config->config_info;
 	DMA_TypeDef *dma = (DMA_TypeDef *)dev_config->base;
+
+	id = id - STREAM_OFFSET;
```
hum, quite ugly to have an expression in the middle of the variable declarations. Well, as you want, it seems that lately declaring variable not at the beginning of a code block is now accepted (it's awful ...).

FRASTM replied:
changed outside the variable declaration

tbursztyka reviewed (commented):
Fix the checkpatch issues.

FRASTM commented:
put `id = id - STREAM_OFFSET` outside the variable declaration

FRASTM committed:
drivers: spi: Enable dma transfer for SPI on stm32

Enable dma operations with or w/o a dmamux on STM32
for SPI periph/memory operations.
Use the pi dma client with dma macros

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
changed the DTS for spi-fifo compatible which requires the dma client definition too

FRASTM commented:
rebase

KwonTae-young commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -451,11 +634,75 @@ static int transceive(struct device *dev,
+	}
+	if (!rx_bufs) {
+		return 0;
+	}
```
Using `spi_read()` and `spi_write()` will return here.
Is only `spi_transceive()` supported?

FRASTM committed:
tests: drivers: spi loopback test application

This adds some commments to the spi_loopback test application

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: drivers: spi loopback test application for dma on stm32

This changes the spi_loopback test application to use the DMA transfer.
In case of DMA, some testcases are useless.
It requires the MOSI to be connected to the MISO of the SPI1.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -24,6 +24,18 @@
+	LL_DMA_PDATAALIGN_HALFWORD,
+	LL_DMA_PDATAALIGN_WORD,
+};
+
```
var `table_m_size` and `table_p_size` are used in the` dma_stm32.c`  in the `dma_stm32_width_config` function.
So it could be defined locally in the function it self

erwango replied:
So maybe this could be done in another commit.
Also, following line in current commit message is not clear to me:
`Fix error of uninitialized values memory_addr_adj, periph_addr_adj` Isn't it linked to next commit (2ceb4ee1aba8ac4888fdf5791f5319f1bba0915d)?

FRASTM replied:
this part is now in the drivers: dma: calculate the data size of the dma_stm32 commit

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -304,6 +325,10 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
+	default:
+		LOG_ERR("Channel direction error. %d",
+				config->channel_direction);
+		return -EINVAL;
```
now put in the commit 2ceb4ee1aba8ac4888fdf5791f5319f1bba0915d

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -68,19 +80,22 @@ static void dma_stm32_irq_handler(void *arg)
-
-		stream->dma_callback(stream->callback_arg, id, 0);
+		stream->dma_callback(stream->callback_arg, id + STREAM_OFFSET,
+					0);
```
set to
>  			0

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -68,19 +80,22 @@ static void dma_stm32_irq_handler(void *arg)
 		LOG_ERR("Unexpected irq happened.");
-		stream->dma_callback(stream->callback_arg, id, -EIO);
+		stream->dma_callback(stream->callback_arg, id + STREAM_OFFSET,
+					-EIO);
```
set to
>  			-EIO

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -63,6 +63,22 @@ static void dma_stm32_irq_handler(void *arg)
+	}
+	/* do not clear the global interrupt yet,
+	 * first handle the specific ones : te or tc
+	 */
```
moved below

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -312,6 +312,16 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
 
+	if ((&config->head_block->source_address == NULL)
+	     || (!config->head_block->source_address)) {
+		LOG_INF("source_buffer address is null.");
```
changed

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -312,6 +312,16 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
+
+	if ((&config->head_block->dest_address == NULL)
+	     || (!config->head_block->dest_address)) {
+		LOG_INF("dest_buffer address is null.");
```
changed to LOG_WRN

FRASTM commented on include/dt-bindings/dma/stm32_dma.h:
```diff
@@ -1,16 +1,38 @@
 /*
  * Copyright (c) 2019 Song Qiang <songqiang1304521@gmail.com>
+ * Copyright (c) 2020 STMicroelectronics
```
commit message changed

FRASTM commented on dts/bindings/spi/st,stm32-spi.yaml:
```diff
@@ -13,3 +13,32 @@ properties:
+      type: string-array
+      description: name of each dma instance
+
+# example for stm32f411 DMA V1 on DMA2 instance
+# for Tx using stream 5 on channel 3 (stream 2 on channel 2 is also possible)
+# for Rx using stream 2 on channel 3 (stream 0 on channel 3 is also possible)
+# &spi1 {
+#	dmas = <&dma2 5 3 0x28440 0x03
+#           &dma2 2 3 0x28480 0x03>;
+#	dma-names = "tx", "rx";
+# };
+#
+# example for stm32l476 DMA V2 on DMA1 instance
+# for Tx using channel 3 with request 1
+# for Rx using channel 2 with request 1
+#&spi1 {
+#	compatible = "st,stm32-spi";
+#	dmas = <&dma1 3 1 0x20440 0x04
+#           &dma1 2 1 0x20480 0x04>;
+#	dma-names = "tx", "rx";
+#};
```
However this explains the dts of the dma client

erwango replied:
`> However this explains the dts of the dma client which is the spi.

Indeed, but this is valid for all dma clients, so should be factorized into dma bindings.
`dma.yaml` would be ideal but, it is not stm32 specific. Only remains `st,stm32-dma.yaml`

FRASTM replied:
example of the client part is given in the `st,stm32-dma.yaml`

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -84,19 +103,26 @@ static void dma_stm32_irq_handler(void *arg)
 		func_ll_clear_tc[id](dma);
 		stream->dma_callback(stream->callback_arg, id + STREAM_OFFSET,
-					0);
+			0);
```
To be moved to commit: `drivers: dma: configuration of the dma_stm32`
But yet the alignment is not correct, this could come from you editor setting (1 tab equals 8 spaces)

FRASTM replied:
moved to commit` drivers: dma: channel configuration for dma_stm32`

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -84,19 +103,26 @@ static void dma_stm32_irq_handler(void *arg)
 	} else if (stm32_dma_is_unexpected_irq_happened(dma, id)) {
 		LOG_ERR("Unexpected irq happened.");
 		stream->dma_callback(stream->callback_arg, id + STREAM_OFFSET,
-					-EIO);
```
To be moved to commit: drivers: dma: configuration of the dma_stm32
But yet the alignment is not correct, this could come from you editor setting (1 tab equals 8 spaces)

FRASTM replied:
moved to commit` drivers: dma: channel configuration for dma_stm32`

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -84,19 +103,26 @@ static void dma_stm32_irq_handler(void *arg)
 		dma_stm32_clear_stream_irq(dev, id);
 
 		stream->dma_callback(stream->callback_arg, id + STREAM_OFFSET,
-					-EIO);
```
To be moved to commit: drivers: dma: configuration of the dma_stm32
But yet the alignment is not correct, this could come from you editor setting (1 tab equals 8 spaces)

FRASTM replied:
moved to a07f54a2f2d76d1c286b46ed172a7d587c534e16

FRASTM commented:
push to consider review and rebase
Add another commit 7f6722078c1a3cbfc9fc24a2f5a4dd29c0e61fa8
Change the commit msg a07f54a2f2d76d1c286b46ed172a7d587c534e16

FRASTM commented:
change dts bindings with an example of the client part is given in the st,stm32-dma.yaml

FRASTM commented on dts/bindings/spi/st,stm32-spi-fifo.yaml:
```diff
@@ -13,3 +13,15 @@ properties:
     interrupts:
       required: true
+
+    dmas:
```
_dmas property in dma.yaml should be changed to required: false_   --> this is not a matter of stm32

FRASTM replied:
this is now done in the commit `dts: bindings: define the bindings for SPI dma client on stm32 series `

FRASTM commented:
changed dt bindings for spi, including the dma.yaml
Consequently the dma.yaml has _dmas_ property '_not required_'

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -519,6 +784,39 @@ static void spi_stm32_irq_config_func_##id(struct device *dev)		\
 #define STM32_SPI_IRQ_HANDLER(id)
 #endif
 
+#ifdef CONFIG_SPI_STM32_DMA
```
Need to do this with DT macros

FRASTM commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -451,11 +634,72 @@ static int transceive(struct device *dev,
+	SPI_TypeDef *spi = cfg->spi;
+	int ret;
+
+	if (!tx_bufs && !rx_bufs) {
```
same a `transceive` function, requires a valid tx and rx buf

tbursztyka replied:
You mean it requires at least one of tx_bufs or rx_bufs so be valid.

erwango commented:
@FRASTM , can you rebase and make use of https://github.com/zephyrproject-rtos/zephyr/pull/24204?

FRASTM commented:
simplify the wrting of macro in the dt-bindings for stm32_dma.h
with 
#define STM32_DMA_CONFIG_PERIPHERAL_DATA_SIZE(config)	(1 << ((config >> 11) & 0x3))
#define STM32_DMA_CONFIG_MEMORY_DATA_SIZE(config)	(1 << ((config >> 13) & 0x3))

avisconti commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -534,6 +829,8 @@ static const struct spi_stm32_config spi_stm32_cfg_##id = {		\
 	SPI_CONTEXT_INIT_LOCK(spi_stm32_dev_data_##id, ctx),		\
 	SPI_CONTEXT_INIT_SYNC(spi_stm32_dev_data_##id, ctx),		\
+	SPI_DMA_CHANNEL_INIT(id, rx, RX, PERIPH, MEM)			\
+	SPI_DMA_CHANNEL_INIT(id, tx, TX, MEM, PERIPH)			\
```
In order to compile properly I had to use PERIPHERAL and MEMORY.

avisconti replied:
ok @FRASTM ,

I see that you fixed the defines instead.
OK, good, I'll test it!

FRASTM commented:
simplify control of source and dest null buffer in commit 8e8cd8b97e7841f32b92f9fd376a455447c28bf1

FRASTM committed:
dts: bindings: define the bindings for SPI dma client on stm32 series

This includes the dma cells as client for the spi
An example of the client part is defined in the dma.yaml
Consequently, bindings for the dmas property becomes then 'not required'
Including dma.yaml is not needed as dma properties
are now part of base.yaml

Signed-off-by: Francois Ramu <francois.ramu@st.com>

avisconti reviewed (approved):
Tested with x_nucleo_iks02a1 shield, microphone sample.
It works fine.
LGTM

FRASTM committed:
tests: drivers: spi loopback testcase to reload dma on stm32

This adds a new testcase in the spi_loopback test application
to use several buffers for the DMA transfer.
It requires the MOSI to be connected to the MISO of the SPI1.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
The dma callback function is now controlling the tx and rx buffers before setting the completion.
Transferring several buffers is possible on the same dma channel with reloading the source (tx) or dest (rx) memory address. 
The tests/drivers/spi/spi_loopback application gives an example `spi_complete_multiple` of this usecase.

FRASTM committed:
drivers: spi: control of the dma callback on spi client for stm32

The DMA callback function now controls the tx or rx buffers
and reload dma in case of multiple trnasfer before the transfer ends

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
checkpatch on the dma_callback function

FRASTM committed:
drivers: dma: stm32 dma_reload function must re-enable

This patch is disabling the dma channel before reloading
source and dest for memory and peripheral addresses
Then the channel is enabled to re-launch the transfer.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on dts/bindings/dma/dma.yaml:
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2019, Linaro Limited
```
This binding is not needed as following properties are now part of base.yaml and hence available to all bindings.

erwango commented on dts/bindings/spi/st,stm32-spi-fifo.yaml:
```diff
@@ -5,11 +5,13 @@ description: STM32 SPI controller with embedded Rx and Tx FIFOs
 compatible: "st,stm32-spi-fifo"
 
-include: spi-controller.yaml
+include: [spi-controller.yaml, dma.yaml]
```
This is not needed as dma properties are now part of base.yaml

FRASTM replied:
removing

erwango commented on dts/bindings/spi/st,stm32-spi.yaml:
```diff
@@ -5,11 +5,13 @@ description: STM32 SPI controller
 compatible: "st,stm32-spi"
 
-include: spi-controller.yaml
+include: [spi-controller.yaml, dma.yaml]
```
This is not needed as dma properties are now part of base.yaml

FRASTM replied:
removing

erwango commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -531,6 +602,11 @@ void test_spi_loopback(void)
 	LOG_INF("SPI test slow config");
 
+	if (spi_complete_multiple(spi_slow, &spi_cfg_slow)
+	    ) {
```
Extra line break to be removed.

FRASTM replied:
done, adding this testcase if  CONFIG_SPI_STM32_DMA

FRASTM commented:
select the testcase for spi_complete_multiple only with SPI on DMA

FRASTM commented:
change `drivers/spi/spi_ll_stm32.c`  to use DT macro

FRASTM commented:
Removing comment in the dts bindings for SPI, information about configuring the DTS for the client is logged `in the st,stm32-dma.yaml`.
The dts_fixup.h is not modified as macro are now defined by the DTS.

erwango commented:
@FRASTM , I propose to close this one and focus efforts on https://github.com/zephyrproject-rtos/zephyr/pull/23396

erwango commented on drivers/spi/Kconfig.stm32:
```diff
@@ -17,10 +17,16 @@ config SPI_STM32_INTERRUPT
 
+config SPI_STM32_DMA
+	bool "STM32 MCU SPI DMA Support"
+	help
```
depends on DMA

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -348,6 +572,14 @@ static int spi_stm32_configure(struct device *dev,
 	ll_func_set_fifo_threshold_8bit(spi);
 #endif
 
+	/* with LL_SPI_FULL_DUPLEX mode, both tx and Rx DMA are on */
```
Shouldn't this be under #idef CONFIG_SPI_STM32_DMA?

FRASTM replied:
yes, changed as the data->dev_dma_tx and data->dev_dma_rx are related to dma too

erwango commented on drivers/spi/Kconfig.stm32:
```diff
@@ -17,10 +17,16 @@ config SPI_STM32_INTERRUPT
 	  Enable Interrupt support for the SPI Driver of STM32 family.
 
+config SPI_STM32_DMA
+	bool "STM32 MCU SPI DMA Support"
```
In I2S driver, which only runs with DMA, enabling I2S selects DMA.
I think we should do the same here, so add:
`select DMA`

FRASTM replied:
added

erwango commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -19,8 +19,28 @@ struct spi_stm32_config {
+
 struct spi_stm32_data {
 	struct spi_context ctx;
+	struct device *dev_dma_tx;
+	struct device *dev_dma_rx;
```
Was there a reason not putting these under CONFIG_SPI_STM32_DMA?

FRASTM replied:
none
changed

erwango commented on drivers/spi/Kconfig.stm32:
```diff
@@ -17,10 +17,16 @@ config SPI_STM32_INTERRUPT
+	bool "STM32 MCU SPI DMA Support"
+	help
+	  Enable DMA support for the SPI Driver of STM32 family.
+
```
Please add some usage guidance such that: when enabled, SPI DMA mode is used for SPI instances that enable dma channels their device tree node.

FRASTM replied:
done

erwango reviewed (changes requested):
Some changes requested.

Also, for tests:
- Remove sentence in commit message about MOSI/MISO connection on SPI1, which does not relates strictly to the commits (this is not a constraint added by the chnage) and is not true (does not have to be SPI1)
- Test assumes that when CONFIG_SPI_STM32_DMA is enabled, test is run on a spi node with dma channels enabled. It fails in some conditions (CONFIG_SPI_ASYNC=y) if run on a node w/o dlma channels enabled. This should be stated in the test (and possibility sanitized txs to DT macros, but this could be done later)

FRASTM commented:
decision to report all tests and configurations in this PR 
(and close PR https://github.com/zephyrproject-rtos/zephyr/pull/23396)

erwango commented on tests/drivers/spi/spi_loopback/boards/nucleo_l476rg.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_SPI_LOOPBACK_DRV_NAME="SPI_1"
+CONFIG_DMA=y
```
This is not needed anymore

FRASTM replied:
removed

erwango replied:
Still there...

erwango commented on tests/drivers/spi/spi_loopback/boards/nucleo_f411re.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_SPI_LOOPBACK_DRV_NAME="SPI_1"
+CONFIG_DMA=y
```
This line is not needed anymore

FRASTM replied:
removed

erwango replied:
Still there

tbursztyka commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -451,11 +624,72 @@ static int transceive(struct device *dev,
+	if ((data->dma_tx.dma_name != NULL)
+	 && (data->dma_rx.dma_name != NULL)) {
+		return transceive_dma(dev, config, tx_bufs, rx_bufs,
+				false, NULL);
```
indentation, it's below last '(' + 1 space (if you configure your editor to follow linux code style it will do it for you)

FRASTM replied:
ok

tbursztyka commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -531,6 +602,11 @@ void test_spi_loopback(void)
 
 	LOG_INF("SPI test slow config");
 
+	if (IS_ENABLED(CONFIG_SPI_STM32_DMA) &&
```
no need to enable this only for STM32_DMA. Looks like a valid test for all soc imo

FRASTM replied:
ok So I add it to the list for slow and fast spi config

FRASTM commented:
removing CONFIG_DMA in the conf of target boards as it is set by the Kconfig of the client

FRASTM commented:
checkpatch: reduce long lines

galak commented on drivers/spi/Kconfig.stm32:
```diff
@@ -17,10 +17,18 @@ config SPI_STM32_INTERRUPT
 	  Enable Interrupt support for the SPI Driver of STM32 family.
 
+config SPI_STM32_DMA
+	bool "STM32 MCU SPI DMA Support"
```
do we need this or could it be driven by having dma property in dts for spi nodes?

erwango replied:
For now, we preferred to keep this way of controlling the feature on top of dts (similarly to SPI_STM32_INTERRUPT). This allows to test both configurations w/o need to change board dts description. I think we'll see in usage if going fully dts driven is more convenient.

galak commented on tests/drivers/spi/spi_loopback/boards/nucleo_f411re.overlay:
```diff
@@ -0,0 +1,19 @@
+ */
+
+&spi1 {
+	dmas = <&dma2 5 3 0x28440 0x03
```
we need a set of macro's/defines to make the magic numbers easier to read (can be a follow up PR).

galak reviewed (commented):
changes look fine, would like to get @tbursztyka, @MaureenHelm or someone from the nordic side to review at the spi loopback test changes.

galak commented:
> changes look fine, would like to get @tbursztyka, @MaureenHelm or someone from the nordic side to review at the spi loopback test changes.

Ah, ignore this bit, see that @tbursztyka already approved this.
